Documentation update for minMaxLoc

refmitchell opened the pull request:
Fixes #25784 

Update documentation for minMaxLoc to be more specific about when multi-channel images are and are not supported.

Testing:
Built documentation locally to check that updates were incorporated correctly.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

dan-masek commented:
IMHO this documentation edit should be a bit more extensive, no need to be so terse.

It would be useful provide an explicit instruction on how to use it to get min/max only on multichannel images.

Furthermore, the documentation should also explicitly mention that this can't work in the Python bindings, given how those output parameters are handled (i.e. there's currently no way to call it from Python while providing the two null pointers necessary for it to work with multi-channel images). That was the main source of confusion in the SO question, if I recall correctly. "Why doesn't this function work the same way in Python as it does in C++?"

refmitchell commented:
Have added more detail about the differences between Python and C++ usage. For 'how to use it' I included a textual description rather than a call signature. As for my original confusion on SO, I'm not too bothered by 'why' things don't work the same way in C++ and Python. I don't feel I need to know the details which lead these two things to behave differently, I just need to know that they behave differently.

**Tangentially related**
I don't know anything about the binding generation process but it feels like these behavioural differences could be eliminated by falling back on C++ defaults if python arguments aren't provided. I'm guessing this is either impractical or impossible? I'm imagining a simple pattern for optional arguments like `int val = python_arg_provided ? python_arg : cpp_default;` but it's probably not that easy.

I also don't know how general a problem this is and how worthwhile it'd be investing the time. Happy to be re-directed if there's a relevant discussion elsewhere.

dan-masek commented:
That's looking much better. Sorry but few small nitpicks.

- `NULL` in context of modern C++ should really be a `nullptr`, don't you think?
- "Or you may extract the particular channel..." -- I was taught that starting a sentence with "or" is at best a little clumsy. A better phrasing could be something like "Another option is to extract the particular channel..."
- "to operate each channel individually" -- you're missing a preposition there IMHO: "to operate on each channel individually"
- "numpy" -- [seems](https://numpy.org/) like the canonical way to write that is "NumPy"

Other than those, looks good to me 👍 

----

"falling back on C++ defaults if python arguments aren't provided"

Therein lies the problem. Let me dump the function signatures here to make it clear:

C++:
`void cv::minMaxLoc(InputArray src,
	double * minVal,
	double * maxVal = 0,
	Point * minLoc = 0,
	Point * maxLoc = 0,
	InputArray mask = noArray())`

Python:
`cv.minMaxLoc(src[, mask]) -> minVal, maxVal, minLoc, maxLoc`

What are "output" arguments (pointers) in the C++ API, are not arguments at all in Python API, they are pure return values. At least the way the bindings are generated right now. From my limited understanding of how those are generated, adding a Python-sensible way to handle that would involve a hand-written wrapper function. Some insight from the maintainers here would be useful. However that is handled, IMHO it should be separate from this task of making the documentation better.

refmitchell commented:
**NULL -> nullptr** 

I don't know enough to argue for or against, it seems fine to update and fine to leave. [Is NULL == nullptr anyway?](https://en.cppreference.com/w/cpp/types/NULL) There are cases where the behaviour of NULL and nullptr can be different but this doesn't seem to be one of them. 

I think a good reason not to change this is that it conflicts with a lot of the other OpenCV documentation (or a least introduces an opportunity for confusion, look at some of the other entries [here](https://docs.opencv.org/4.x/d2/de8/group__core__array.html)). Updating this to be more inline with modern C++ feels like a more general (separate) issue. (Here we should use NULL to remain consistent, then a separate issue should be opened to bring the docs up to date more generally.) 

Let me know your thoughts on this and I'll push an update (along with the other fixes mentioned below).

**Other stuff**

I've addressed the other stuff you mentioned as follows:

- Starting with 'Or': This was part of the original entry which I was trying to leave as intact as possible (in order to avoid introducing any other errors).  I've changed this to 'Alternatively, you can...' as I think it reads better.  That said, the 'or' phrasing appears in a few other entries in the docs. (Go to [the operations on arrays page](https://docs.opencv.org/4.x/d2/de8/group__core__array.html) and search for "The function do not work" [yes, it is grammatically incorrect].) As an aside, starting sentences with conjunctions is fine in English. Whether or not you choose to do this is contextual/stylistic - it depends on how you want it to read. It isn't inherently wrong or clumsy.
- Missing preposition: fixed.
- numpy -> NumPy: fixed.


**Default arguments**

I was just curious about the binding generation process, I agree it's a separate issue and we don't need to discuss it further here.

dan-masek commented:
Makes sense, and to me the current state looks good enough to move on.

Thank you.

dan-masek commented:
@asmorkalov Strange that those builds failed. Looks like a pure documentation change, and those failures were somewhere in dnn module? Any way to restart them easily?

refmitchell commented:
Do we have any idea why these builds are failing? Is there anything I have done wrong or can do to fix?

sturkmen72 commented:
@refmitchell "Error: The action 'Build OpenCV' has timed out after 60 minutes." the fails are not related with the changes you have done.

refmitchell commented:
I thought so but wanted to check in with someone who knows how this build is put together.
